### PR TITLE
[W3M] Expose modal Web3ModalComponent and ComponentDelegate

### PR DESCRIPTION
**Pull request by kacperoak**
- Expose Web3ModalComponent for developers
- Add component delegate

### CODE REVIEW

File: product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/ComponentDelegate.kt [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.walletconnect.web3.modal.ui.components
+
+import com.walletconnect.web3.modal.client.Web3Modal
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.onEach
+
+internal sealed class ComponentEvent {
+    object ModalHiddenEvent: ComponentEvent()
+    object ModalExpandedEvent: ComponentEvent()
+}
```
Comment (TalhaAli00) [inline product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/ComponentDelegate.kt:10]: In the Web3ModalState file, you check each of these values and return a boolean. 

You could also do
```suggestion
internal sealed class ComponentEvent(isOpen: Boolean) {
    object ModalHiddenEvent: ComponentEvent(false)
    object ModalExpandedEvent: ComponentEvent(true)
}
```

Then in Web3ModalSate where you listen to the event, you just grab the `isOpen` variable

File: product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/button/Web3ModalState.kt [hunk 2]
```diff
@@ -41,12 +39,20 @@ class Web3ModalState(
...
     private val getSelectedChainUseCase: GetSelectedChainUseCase = wcKoinApp.koin.get()
 
-    val isOpen = navController.currentBackStackEntryFlow.mapLatest { it.destination.route?.startsWith(Route.WEB3MODAL.path) ?: false }
+    val isOpen = ComponentDelegate.modalComponentEvent
+        .map { event ->
+            when (event) {
+                ComponentEvent.ModalHiddenEvent -> false
+                ComponentEvent.ModalExpandedEvent -> true
+            }
+        }
```
Comment (TalhaAli00) [inline product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/button/Web3ModalState.kt:54]: If you want to use the suggestion above, then this becomes

```suggestion
    val isOpen = ComponentDelegate.modalComponentEvent
        .map { event -> event.isOpen }
```

**Review by TalhaAli00 — APPROVED** [review]

File: product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/ComponentDelegate.kt (hunk 1 above)
Comment (jakubuid) [reply product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/ComponentDelegate.kt:10]: Why would we need it? I think just by an event itself it's clear on which state modal is

Comment (jakubuid) [reply product/web3modal/src/main/kotlin/com/walletconnect/web3/modal/ui/components/ComponentDelegate.kt:10]: Ah ok I see below

**Review by jakubuid — APPROVED** [review]

**Pull request merged**